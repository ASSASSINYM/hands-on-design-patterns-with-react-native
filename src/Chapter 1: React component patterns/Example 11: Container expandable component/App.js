import React from 'react';
import {StyleSheet, Text, View} from 'react-native';

// Tip: Below components could be placed in separate files.
//      One file per one component.

export const HelloText = ({children, ...otherProps}) => (
    <Text {...otherProps}>{children}</Text>
);

export const HelloBox = ({
    isExpanded,
    expandOrCollapse,
    containerStyles,
    expandedTextStyles
}) => (
    <View style={containerStyles || styles.container}>
        <HelloText onPress={() => expandOrCollapse()}>
            Hands-On Design Patterns with React Native
        </HelloText>
        <HelloText onPress={() => expandOrCollapse()}>
            Chapter 1: React Component Patterns
        </HelloText>
        {
            isExpanded &&
            <HelloText style={expandedTextStyles || styles.text}>
                You can expand and collapse this text by clicking
                the Title or Chapter text. Bonus: Check Chapter 4
                to learn how to animate expanding andcollapsing.
            </HelloText>
        }
    </View>
);

export default class App extends React.Component {
    constructor() {
        super();
        this.state = {
            // default state on first render
            expanded: false
        };
        this.expandOrCollapse = this.expandOrCollapse.bind(this);
    }

    expandOrCollapse() {
        // toggle expanded: true becomes false, false becomes true
        this.setState({expanded: !this.state.expanded});
    }

    render = () => (
        <HelloBox
            isExpanded={this.state.expanded}
            expandOrCollapse={this.expandOrCollapse}
            expandedTextStyles={{ color: 'red' }}
        />
    );
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor: '#fff',
        alignItems: 'center',
        padding: '10%',
        justifyContent: 'center',
    },
    text: {
        marginTop: '5%',
    }
});
