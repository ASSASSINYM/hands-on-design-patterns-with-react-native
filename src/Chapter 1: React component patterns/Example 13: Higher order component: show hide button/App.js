import React from 'react';
import {StyleSheet, Text, View, Button} from 'react-native';

export const SomeSection = ({
    isExpanded,
    expandOrCollapse,
    containerStyles,
    boxStyle
}) => (
    <View style={containerStyles || styles.container}>
        <Button
            onPress={expandOrCollapse}
            title={isExpanded ? "Hide" : "Show"}
            color="#841584"
        />
        {isExpanded && <View style={boxStyle || styles.box}/>}
    </View>
);

const makeExpandable = (ComponentToEnrich) => (
    class HelloBoxContainer extends React.Component {
        constructor() {
            super();
            this.state = {
                // default state on first render
                expanded: false
            };
            this.expandOrCollapse = this.expandOrCollapse.bind(this);
        }

        expandOrCollapse() {
            // toggle expanded: true becomes false, false becomes true
            this.setState({expanded: !this.state.expanded});
        }

        render = () => (
            <ComponentToEnrich
                isExpanded={this.state.expanded}
                expandOrCollapse={this.expandOrCollapse}
            />
        );
    }
);

export default makeExpandable(SomeSection);

const styles = StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor: '#fff',
        alignItems: 'center',
        padding: '10%',
        justifyContent: 'center',
    },
    box: {
        width: 100,
        height: 100,
        backgroundColor: 'powderblue',
    }
});
