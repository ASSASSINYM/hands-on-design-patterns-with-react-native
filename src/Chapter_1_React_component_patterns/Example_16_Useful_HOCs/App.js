import React from 'react';
import {TouchableHighlight, StyleSheet, Text, View, Button, ActivityIndicator, Modal} from 'react-native';

export const SomeSection = ({
    showHideBox,
    isVisible,
    containerStyles,
    boxStyle
}) => (
    <View style={containerStyles || styles.container}>
        <Button
            onPress={showHideBox}
            title={isVisible ? "Hide" : "Show"}
            color="#841584"
        />
        {isVisible && <View style={boxStyle || styles.box}/>}
    </View>
);

const makeExpandable = (ComponentToEnrich) => (
    class HelloBoxContainer extends React.Component {
        constructor() {
            super();
            this.state = {
                // default state on first render
                expanded: false
            };
            this.expandOrCollapse = this.expandOrCollapse.bind(this);
        }

        expandOrCollapse() {
            // toggle expanded: true becomes false, false becomes true
            this.setState({expanded: !this.state.expanded});
        }

        render = () => (
            <ComponentToEnrich
                isExpanded={this.state.expanded}
                expandOrCollapse={this.expandOrCollapse}
            />
        );
    }
);

const mapPropNames = (Component) => (props) => (
    <Component
        {...props}
        isVisible={props.isExpanded}
        showHideBox={props.expandOrCollapse}
    />
);

const logPropChanges = Component => props => {
    console.log('[Actual props]:', props);
    return <Component {...props} />;
};

const withSpinner = Component => props => (
    // Needs a HOC that provides prop shouldSpin (true/false)
    // Try setting it yourself
    props.shouldSpin
        ? <View style={styles.container}>
            <Text>Your fav spinner f.in. on data load.</Text>
            <ActivityIndicator size="large" color="#0000ff" />
        </View>
        : <Component {...props} />
);

const withModalOpener = OriginalComponent => (
    class ModalExample extends React.Component {
        state = {
            modalVisible: true,
        };

        setModalVisible(visible) {
            this.setState({modalVisible: visible});
        }

        render() {
            return (
                // Replace with your favourite Modal box implementation
                <View style={styles.container}>
                    <OriginalComponent
                        {...this.props}
                        openModal={() => this.setModalVisible(true)}
                        closeModal={() => this.setModalVisible(false)}
                    />
                    <Modal
                        animationType="slide"
                        visible={this.state.modalVisible}
                        onRequestClose={() => {
                            alert('Modal has been closed.');
                        }}>
                        <View style={styles.container}>
                            <Text>Example modal!</Text>

                            <TouchableHighlight
                                onPress={() => {
                                    this.setModalVisible(false);
                                }}>
                                <Text style={{fontSize: 30}}>
                                    Hide Modal
                                </Text>
                            </TouchableHighlight>
                        </View>
                    </Modal>
                </View>
            );
        }
    }
);

export default withModalOpener(makeExpandable(logPropChanges(mapPropNames(SomeSection))));

const styles = StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor: '#fff',
        alignItems: 'center',
        padding: '10%',
        justifyContent: 'center',
    },
    box: {
        width: 100,
        height: 100,
        backgroundColor: 'powderblue',
    }
});
