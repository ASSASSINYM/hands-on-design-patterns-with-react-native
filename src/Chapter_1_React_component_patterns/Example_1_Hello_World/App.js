import React from 'react';
import {StyleSheet, Text, View} from 'react-native';

export default class App extends React.Component {
    render = () => (
        <View style={styles.container}>
            <Text>Hands-On Design Patterns with React Native</Text>
            <Text>Chapter 1: React Component Patterns</Text>
            <Text style={styles.text}>
                You are ready to start the journey. Fun fact is, this
                text is rendered by class component called App. Check
                App.js if you want to look it up.
            </Text>
        </View>
    );
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor: '#fff',
        alignItems: 'center',
        padding: '10%',
        justifyContent: 'center',
    },
    text: {
        marginTop: '5%',
    }
});
