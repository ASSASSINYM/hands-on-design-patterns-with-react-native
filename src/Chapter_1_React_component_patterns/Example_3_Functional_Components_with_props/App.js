import React from 'react';
import {StyleSheet, Text, View} from 'react-native';

const HelloText = ({children, ...otherProps}) => (
    <Text {...otherProps}>{children}</Text>
);
const App = () => (
    <View style={styles.container}>
        <HelloText>
            Hands-On Design Patterns with React Native
        </HelloText>
        <HelloText>Chapter 1: React Component Patterns</HelloText>
        <HelloText style={styles.text}>
            You are ready to start the journey. Fun fact is, this text
            is rendered by Functional Component called HelloText.
            Check App.js if you want to look it up.
        </HelloText>
    </View>
);
export default App;

const styles = StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor: '#fff',
        alignItems: 'center',
        padding: '10%',
        justifyContent: 'center',
    },
    text: {
        marginTop: '5%',
    }
});
