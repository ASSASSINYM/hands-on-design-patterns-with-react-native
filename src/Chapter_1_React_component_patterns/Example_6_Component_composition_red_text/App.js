import React from 'react';
import {StyleSheet, Text, View} from 'react-native';

const WarningText = ({style, ...otherProps}) => (
    <Text style={[style, {color: 'orange'}]} {...otherProps} />
);

export default class App extends React.Component {
    render = () => (
        <View style={styles.container}>
            <Text style={styles.text}>Normal text</Text>
            <WarningText style={styles.text}>Warning</WarningText>
        </View>
    );
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor: '#fff',
        alignItems: 'flex-start',
        padding: '10%',
        justifyContent: 'flex-start',
        width: '100%'
    },
    text: {
        marginTop: '5%',
        fontWeight: 'bold',
        fontSize: 40,
    }
});
