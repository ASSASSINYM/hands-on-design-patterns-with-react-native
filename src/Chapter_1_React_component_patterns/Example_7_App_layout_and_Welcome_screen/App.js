import React from 'react';
import {StyleSheet, Text, View} from 'react-native';

const AppLayout = ({header, mainContent, footer}) => (
    <View style={styles.container}>
        <View style={styles.layoutHeader}>{header}</View>
        <View style={styles.layoutContent}>{mainContent}</View>
        <View style={styles.layoutFooter}>{footer}</View>
    </View>
);

const WelcomeHeader = () => <View><Text>Header</Text></View>;
const WelcomeContent = () => <View><Text>Content</Text></View>;
const WelcomeFooter = () => <View><Text>Footer</Text></View>;

const WelcomeScreen = () => (
    <AppLayout
        header={<WelcomeHeader />}
        mainContent={<WelcomeContent />}
        footer={<WelcomeFooter />}
    />
);

export default class App extends React.Component {
    render = () => <WelcomeScreen />;
}

// Don't worry if you don't know Flexbox in CSS, we cover it later
const styles = StyleSheet.create({
    container: {
        flex: 1,
        marginTop: 20
    },
    layoutHeader: {
        width: '100%',
        height: 100,
        backgroundColor: 'powderblue'
    },
    layoutContent: {
        flex: 1,
        width: '100%',
        backgroundColor: 'skyblue'
    },
    layoutFooter: {
        width: '100%',
        height: 100,
        backgroundColor: 'steelblue'
    }
});
