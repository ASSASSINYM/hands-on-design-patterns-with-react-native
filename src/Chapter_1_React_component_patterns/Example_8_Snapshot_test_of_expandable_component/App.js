import React from 'react';
import {StyleSheet, Text, View} from 'react-native';

const HelloText = ({ children, ...otherProps }) => (
    <Text {...otherProps}>{children}</Text>
);
export default class App extends React.Component {
    constructor() {
        super();
        this.state = {
            // default state on first render
            expanded: false
        };
    }

    expandOrCollapse() {
        // toggle expanded: true becomes false, false becomes true
        this.setState({expanded: !this.state.expanded});
    }

    render = () => (
        <View style={styles.container}>
            <HelloText onPress={() => this.expandOrCollapse()}>
                Hands-On Design Patterns with React Native
            </HelloText>
            <HelloText onPress={() => this.expandOrCollapse()}>
                Chapter 1: React Component Patterns
            </HelloText>
            {
                this.state.expanded &&
                <HelloText style={styles.text}>
                    You can expand and collapse this text by clicking
                    the Title or Chapter text. Bonus: Check Chapter 4
                    to learn how to animate expanding andcollapsing.
                </HelloText>
            }
        </View>
    );
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor: '#fff',
        alignItems: 'center',
        padding: '10%',
        justifyContent: 'center',
    },
    text: {
        marginTop: '5%',
    }
});
