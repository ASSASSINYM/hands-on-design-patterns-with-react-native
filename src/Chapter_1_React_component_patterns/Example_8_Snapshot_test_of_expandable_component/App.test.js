import React from 'react';
import App from './App';

import renderer from 'react-test-renderer';

it('renders', () => {
    const rendered = renderer.create(<App />).toJSON();
    expect(rendered).toMatchSnapshot();
});
