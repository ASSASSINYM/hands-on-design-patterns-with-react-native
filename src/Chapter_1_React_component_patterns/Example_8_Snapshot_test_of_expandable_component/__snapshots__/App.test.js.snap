// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
      "padding": "10%",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    onPress={[Function]}
  >
    Hands-On Design Patterns with React Native
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    onPress={[Function]}
  >
    Chapter 1: React Component Patterns
  </Text>
</View>
`;
